runtime/pubsub/topic.go

```diff
@@ -24,68 +28,155 @@ type Topic[T any] interface {
+	topicCfg *config.PubsubTopic
+	topic    types.TopicImplementation
 }
 
-// SubscriptionConfig is used when creating a subscription
-type SubscriptionConfig struct {
-	// Filter is a boolean expression using =, !=, IN, &&
-	// It is used to filter which messages are forwarded from the
-	// topic to a subscription
-	Filter string
+func (t *topicAdapter[T]) NewSubscription(name string, sub Subscriber[T], cfg *SubscriptionConfig) Subscription[T] {
+	// Fetch the subscription configuration
+	subscription, ok := t.topicCfg.Subscriptions[name]
+	if !ok {
+		logging.RootLogger.Fatal().Msgf("unregistered/unknown subscription on topic %s: %s", t.topicCfg.EncoreName, name)
+	}
+
+	staticCfg, ok := config.Cfg.Static.PubsubTopics[t.topicCfg.EncoreName].Subscriptions[name]
+	if !ok {
+		logging.RootLogger.Fatal().Msgf("unregistered/unknown subscription on topic %s: %s", t.topicCfg.EncoreName, name)
+	}
+
+	panicCatchWrapper := func(ctx context.Context, msg T) (err error) {
+		defer func() {
+			if err2 := recover(); err2 != nil {
+				err = errs.B().Code(errs.Internal).Msgf("subscriber paniced: %s", err2).Err()
+			}
+		}()
+
+		return sub(ctx, msg)
+	}
+
+	log := logging.RootLogger.With().
+		Str("service", staticCfg.Service.Name).
+		Str("topic", t.topicCfg.EncoreName).
+		Str("subscription", name).
+		Logger()
+
+	// Subscribe to the topic
+	t.topic.Subscribe(&log, cfg, subscription, func(ctx context.Context, msgID string, publishTime time.Time, deliveryAttempt int, attrs map[string]string, data []byte) (err error) {
+		runtime.BeginOperation()
+		defer runtime.FinishOperation()
+
+		var msg T
+
+		if err = json.Unmarshal(data, &msg); err != nil {
+			err = errs.B().Cause(err).Code(errs.InvalidArgument).Msg("failed to unmarshal message").Err()
+			log.Err(err).Str("msg-id", msgID).Int("delivery-attempt", deliveryAttempt).Msg("failed to unmarshal message")
+			return err
+		}
+
+		if err = utils.UnmarshalFields(attrs, &msg, utils.AttrTag); err != nil {
+			err = errs.B().Cause(err).Code(errs.InvalidArgument).Msg("failed to unmarshal attributes").Err()
+			log.Err(err).Str("msg-id", msgID).Int("delivery-attempt", deliveryAttempt).Msg("failed to unmarshal message attributes")
+			return err
+		}
+
+		// Start the request tracing span
+		err = runtime.BeginRequest(ctx, runtime.RequestData{
+			Type:    runtime.PubSubMessage,
+			Service: staticCfg.Service.Name,
+			MsgData: runtime.PubSubMsgData{
+				Topic:        t.topicCfg.EncoreName,
+				Subscription: subscription.EncoreName,
+				MessageID:    msgID,
+				Attempt:      deliveryAttempt,
+				Published:    publishTime,
+			},
+			CallExprIdx:     0,
+			EndpointExprIdx: staticCfg.TraceIdx,
+			Inputs:          [][]byte{data},
+		})
+		if err != nil {
+			return errs.B().Code(errs.Internal).Cause(err).Msg("failed to begin request").Err()
+		}
+
+		err = panicCatchWrapper(ctx, msg)
+		runtime.FinishRequest(nil, err)
+		return err
+	})
+
+	log.Info().Msg("registered subscription")
+
+	return struct{}{}
```

Comment:
How does `struct{}{}` implement `Subscription[T]`? Or am I misunderstanding something?

Reply:
`Subscription[T]` is an empty interface, so the struct matches that need. But it means that these functions are already returning a value, so if we add a method to `Subscription[T]` this will cause a nice compile error

Reply:
Ah, nice :)